shell/run.ts

```diff
@@ -0,0 +1,181 @@
+ *
+ * @param command The {@link ProcessCommand} to be run
+ * @param options The {@link FullUserRunOptions} governing the behavior of the process runner
+ * @returns A {@link Promise<ProcessResult>} containing the result state when the process completes.
+ */
+$.run = runGeneric;
+
+/**
+ * Runs a process using the piped output API.
+ *
+ * Note: This is the same as using {@link $} on its own, except that it provides the full process result instead of
+ * just the string output.
+ *
+ * @param command The {@link ProcessCommand} to be run
+ * @param options The {@link FullUserRunOptions} governing the behavior of the process runner
+ * @returns A {@link Promise<PipedProcessResult>} containing the result state and output when the process completes
+ */
+$.piped = runPiped;
+
+/**
+ * Runs a process using the streamed output API.
+ *
+ * This means that output will be sent to the same `stdout`/`stderr` as the Deno script, so it will show up in along with
+ * console output.
+ *
+ * @param command The {@link ProcessCommand} to be run
+ * @param options The {@link FullUserRunOptions} governing the behavior of the process runner
+ * @returns A {@link Promise<ProcessResult>} containing the result state when the process completes
+ */
+$.streamed = runStreamed;
+
+/**
+ * Runs a process and checks to see if it was successful.
+ *
+ * @param command The {@link ProcessCommand} to be run
+ * @param options The {@link FullUserRunOptions} governing the behavior of the process runner
+ * @returns A {@link Promise<boolean>} that will resolve to `true` if the process had a zero exit code to indicate success
+ */
+$.try = tryRun;
+
+async function tryRun(
+  command: string[] | string,
+  options: UserTryOptions = {},
+): Promise<boolean> {
+  try {
+    const result = await runPiped(command, { ...options, throwOnError: false });
+    return result.status.success;
+  } catch (err) {
+    if (err instanceof Deno.errors.NotFound) {
+      return false;
+    } else {
+      throw err;
+    }
+  }
+}
+
+function runStreamed(
+  command: string[] | string,
+  options: UserRunOptions = {},
+): Promise<ProcessResult> {
+  const fullOptions: FullRunOptions = {
+    cmd: buildCommandArray(command),
+    stdout: 'inherit',
+    stderr: 'inherit',
+    ...options,
+  };
+  return runInternal(fullOptions);
+}
+
+function runPiped(
+  command: string[] | string,
+  options: UserRunOptions = {},
+): Promise<PipedProcessResult> {
+  const fullOptions: PipedRunOptions = {
+    cmd: buildCommandArray(command),
+    stdout: 'piped',
+    stderr: 'piped',
+    ...options,
+  };
+  return runInternal(fullOptions);
+}
+
+function runGeneric(
+  command: string[] | string,
+  options: FullUserRunOptions = {},
+): Promise<ProcessResult> {
+  const fullOptions: FullRunOptions = {
+    cmd: buildCommandArray(command),
+    ...options,
+  };
+  return runInternal(fullOptions);
+}
+
+function runInternal(options: PipedRunOptions): Promise<PipedProcessResult>;
```

Comment:
What are these declarations for?

Reply:
This is just creating a function overload for TypeScript. When the `options` are known to specify piping (`PipedRunOptions` vs `FullRunOptions`), then the returned result will definitely have piped output available (`PipedProcessResult` vs `ProcessResult`).

The last two are identical because TypeScript hides the implementation function from direct calls, but in this case, we need to directly call it. So I added an extra overload definition for that signature to be callable.

https://www.typescriptlang.org/docs/handbook/2/functions.html#function-overloads